WebKit2GTK: Add wxWebViewConfiguration and child window support

TcT2k opened the pull request:
Add some features to the WebKit2GTK backend that where recently added to Edge and macOS

wxWebViewConfiguration
==

Add support for more in depth configuration before creating a web view. Allows to customize the data path and access to the underlying context.

Add wxWebViewWindowFeatures and event
==

This replaces the previously implemented `wxWebViewWindowInfo`.

It explicitly **breaks** the previous API to enable WebKitGTK
integration and to make the usage in application code less error prone.

A new event `wxEVT_WEBVIEW_NEWWINDOW_FEATURES` is added to allow
access to the window features and the child web view.

New sample handling code looks like this:
```C++
// Bind new window handler
m_webView->Bind(wxEVT_WEBVIEW_NEWWINDOW, [](wxWebViewEvent& evt) {
    if (evt.GetURL() == "http://badwebsite.com")
        evt.Veto(); // Disallow new window for badwebsite.com
    else
        evt.Skip(); // Allow new window for all other websites
});
 
// Bind new window features handler
m_webView->Bind(wxEVT_WEBVIEW_NEWWINDOW_FEATURES, [](wxWebViewEvent& evt) {
    // Get target window features
    wxWebViewWindowFeatures* features = evt.GetTargetWindowFeatures();
    // Create a top level window for the child web view
    wxWindow* win = new wxWindow(this, wxID_ANY, features->GetPosition(), features->GetSize());
    wxSizer* sizer = new wxBoxSizer(wxVERTICAL);
    win->SetSizer(sizer);
    // Get the child web view
    wxWebView* childWebView = features->GetChildWebView();
    // Place the child web view into the window
    childWebView->Create(win, wxID_ANY);
    sizer->Add(childWebView, 1, wxEXPAND);
}
```

TcT2k committed:
Implement wxEVT_WEBVIEW_WINDOW_CLOSE_REQUESTED

Implement wxEVT_WEBVIEW_WINDOW_CLOSE_REQUESTED for webkitgtk2.

TcT2k committed:
Unify entries in wxWebViewEvent table

TcT2k committed:
Remove unused GTK WebView new window handling code

TcT2k committed:
Add navigation action for webkitgtk2 backend

vadz commented on include/wx/gtk/webview_webkit.h:
```diff
@@ -216,9 +212,18 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewFactoryWebKit : public wxWebViewFactory
+        if (webView->Create(parent, id, url, pos, size, style, name))
+            return webView;
+        else
+            return nullptr;
```
Shouldn't we `delete webView` here (or, rather, use `std::unique_ptr<>` for it in the first place)?

TcT2k replied:
Changed to use `std::unique_ptr`

vadz commented on src/common/webview.cpp:
```diff
@@ -86,6 +87,24 @@ wxString wxWebViewConfiguration::GetDataPath() const
+
+wxWebViewWindowFeatures::~wxWebViewWindowFeatures()
+{
+    if (m_childWebView && !m_childWebViewWasUsed)
+        delete m_childWebView;
```
I wonder if we could avoid this by making `GetChildWebView()` reset the pointer to `nullptr` after returning it? In fact, what about using `unique_ptr<>` here too and returning it by value from `GetChildWebView()` to make it absolutely clear that the caller takes its ownership?

TcT2k replied:
Using a `std::unique_ptr` would probably make sense. But I intentionally wanted `GetChildWebView()` to be callable multiple times by application code. E.g. it could make sense to do something with the webview before processing the rest of `wxWebViewWindowFeatures` in another method.

As this is a control the application code shouldn't have to worry to much about ownership as that is more or less automatic once it's added to a window.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -608,6 +682,80 @@ wxgtk_authorize_authenticated_peer_cb(GDBusAuthObserver *,
+public:
+    wxWebViewConfigurationImplWebKit():
+        m_webContext(nullptr)
+#if WEBKIT_CHECK_VERSION(2, 10, 0)
```
Should we perhaps define some `HAVE_WEBKIT_WEBSITE_DATA_MANAGER` once instead of checking for the version multiple times? This would be a bit more readable, IMHO, and would make it easier to get rid of this check later when we drop support for the versions < 2.10.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -608,6 +682,80 @@ wxgtk_authorize_authenticated_peer_cb(GDBusAuthObserver *,
+    {
+        GetOrCreateContext();
+        if (m_websiteDataManager)
+            return webkit_website_data_manager_get_base_data_directory(m_websiteDataManager);
```
We probably need a run-time version check for 2.10 here (and when creating it below), otherwise an app compiled on a newer system will crash when running on an older one and I think we want to support this.

As I've just recently written in [this comment](https://github.com/wxWidgets/wxWidgets/issues/23367#issuecomment-1479377777), I think it would be useful to add `wx_check_webkit_version()` function for doing it, as WebKit itself doesn't seem to provide one.

vadz replied:
Yes, it's called "lazy binding" and is the default under Linux (and the vast majority of the other Unix systems although historically there were some exceptions, I believe). This is also the default behaviour of `dlopen()` (`RTLD_LAZY` is the default, but `RTLD_NOW` can be used to force immediately resolving everything).

vadz reviewed (commented):
Thanks for your work on this!

The new API looks perfectly reasonable to me and it's great that it now works on all platforms.

I just have some minor comments about (not) deleting raw pointers, I think it would be nice to start following good C++ practices in wx code (better late than never, right) and use `unique_ptr` instead whenever possible.

Thanks again!

TcT2k commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -73,6 +74,79 @@ bool wxGetStringFromJSResult(WebKitJavascriptResult* js_result, wxString* output
+            m_geometry.width = -1;
+        if (m_geometry.height == 0)
+            m_geometry.height = -1;
+        if (m_geometry.x == 0)
+            m_geometry.x = -1;
+        if (m_geometry.y == 0)
+            m_geometry.y = -1;
```
While valid I think it's still the right thing to do. As in most cases it won't be set and it's more in line with the other backends (where we can actually check if a value is available).

TcT2k committed:
Add wxWebViewWindowFeatures and event

This replaces the previously implemented wxWebViewWindowInfo.

It explicitly breaks the previous API to enable WebKitGTK
integration and to make the usage in application code less error prone.

A new event wxEVT_WEBVIEW_NEWWINDOW_FEATURES is added to allow
access to the window features and the child web view.

vadz commented on include/wx/webview.h:
```diff
@@ -361,10 +365,11 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewWindowInfo
 
-    virtual wxWebView* CreateChildWebView() = 0;
+protected:
+    mutable bool m_childWebViewWasUsed; // modified from const method
```
Sorry for continuing to nitpick, but `GetChildWebView()` is not actually const, which is why I was wondering why it was mutable in the first place. Did you intend to make it const, or is there some other method modifying this variable somewhere?

TcT2k replied:
I see, I indeed intented it to be const. Will change that.

swt2c commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -38,6 +39,20 @@
+                                 webkit_get_micro_version();
+    const unsigned int required = major * 10000 + minor * 100 + micro;
+    return version >= required;
+}
```
Can `wx_check_webkit_version` please be moved into a separate commit, and perhaps be merged separately ASAP?  I could use it for the other webkit2 changes.

TcT2k replied:
I've added it as a separate commit. But this PR can probably be merged soon anyway and would include this change.

swt2c replied:
Thanks - that will make it easier to backport - since all of the changes in this PR will probably not be backportable.

TcT2k committed:
Add wx_check_webkit_version()

This function allows for runtime checks of the webkit version

TcT2k committed:
Add wxWebViewConfiguration support for webkitgtk2

Allows setting some webview settings which would be unavailable after
creation.

vadz commented:
Thanks, merged now and, indeed, this won't be backported entirely but we can backport 42ca043.
